### ISSUE TITLE
Any plan to add string iterator?

### ISSUE THREAD

**Issue by windowsair** [simdutf/simdutf#259]
It would be great if there was some way to iterate over something like each UTF8 character. 

Are there plans to add this in the future?

**Comment by lemire** [comment]
Can you elaborate? If you mean iterating over the code point values, we do fast transcoding to UTF-32. After UTF-32 trancoding, iteration is trivial.

For large inputs, it might make sense to do block-wise UTF-32 decoding to avoid overwhelming the cache.

Byte-by-byte processing is inefficient and not something we want to encourage.

**Comment by windowsair** [comment]
Yes, I think what I want is to iterate over code point values or something like that.

```cpp
char buf[] = "Foo © bar 𝌆 baz ☃ qux";

...

for (char8_t x: itor) {
    std::cout << x;  // and we can get each UTF8 character like "©", "𝌆", "☃"
}
```

It is easier to implement a simple iterator. However, trying to take advantage of features such as SIMD may be difficult, for byte-by-byte iteration.

**Comment by lemire** [comment]
Interesting.

**Comment by WojciechMula** [comment]
I think a character-wise iterator can anyway use SIMD backend, just silently transcode bigger chunks of the input. The question is, if in such cases performance is the key. I mean, if you want to analyse the input char-by-char means you need some extra processing of char code itself.

Anyway, we need to define that it yields `char32_t` or `uint32_t` or `uint64_t` (the last one for completeness).

It would be good to take a look how the Go runtime handles iteration over UTF-8.

**Comment by lemire** [comment]
To iterate over code points efficiently, transcode to UTF-32 (in blocks, for large inputs) and iterate over the result. `utf32_length_from_utf8` and `convert_utf8_to_utf32` make that simple, and it will be much faster than decoding one byte at a time. We don't plan to add a dedicated iterator type, so I'm closing this. A higher-level ranges-style API is part of #144.
